pkg/apis/serving/v1beta1/explainer.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2020 kubeflow.org.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// ExactlyOneExplainerViolatedError is a known error message
+	ExactlyOneExplainerViolatedError = "Exactly one of [Custom, Alibi] must be specified in ExplainerSpec"
+)
+
+// ExplainerSpec defines the container spec for a model explanation server,
+// The following fields follow a "1-of" semantic. Users must specify exactly one spec.
+type ExplainerSpec struct {
+	// Spec for alibi explainer
+	Alibi *AlibiExplainerSpec `json:"alibi,omitempty"`
+	// Pass through Pod fields or specify a custom container spec
+	*CustomExplainer `json:",inline"`
+	// Extensions available in all components
+	*ComponentExtensionSpec `json:",inline"`
+}
+
+// Explainer interface is implemented by all explainers
+// +kubebuilder:object:generate=false
+type Explainer interface {
```

Comment:
I think we should be able to abstract this interface into component.go that implements:
```
GetContainer()
Default()
Validate()
```
Predictor/Explainer/Transformer can all implement this.


We can have a second interface 
explainer.go
that has the specialized function
```
GetStorageUri()
```

I think this should clean up a ton of validation/defaulting/retrieval logic.

Reply:
you are right, I have consolidated them into the `Component` interface